pkg/pool/linked_list_test.go

```diff
@@ -0,0 +1,365 @@
+			},
+		},
+		{
+			name: "Works with 1 insert",
+			before: func(dll *DoubleLinkedList[string]) []*Node[string] {
+				nodes := append([]*Node[string]{}, dll.Insert("One"))
+
+				return nodes
+			},
+			apply: func(dll *DoubleLinkedList[string], n []*Node[string]) {
+				for _, node := range n {
+					dll.Delete(node)
+				}
+			},
+			check: func(dll *DoubleLinkedList[string]) bool {
+				if dll.Len() != 0 {
+					return false
+				}
+
+				return reflect.DeepEqual(dll.toArray(), []string{})
+			},
+		},
+		{
+			name: "Works with 2 inserts",
+			before: func(dll *DoubleLinkedList[string]) []*Node[string] {
+				nodes := append([]*Node[string]{}, dll.Insert("One"), dll.Insert("Two"))
+
+				return nodes
+			},
+			apply: func(dll *DoubleLinkedList[string], n []*Node[string]) {
+				for _, node := range n {
+					dll.Delete(node)
+				}
+			},
+			check: func(dll *DoubleLinkedList[string]) bool {
+				if dll.Len() != 0 {
+					return false
+				}
+
+				return reflect.DeepEqual(dll.toArray(), []string{})
+			},
+		},
+		{
+			name: "Works with 100 inserts",
+			before: func(dll *DoubleLinkedList[string]) []*Node[string] {
+				nodes := []*Node[string]{}
+
+				for i := 0; i < 100; i++ {
+					nodes = append(nodes, dll.Insert("Test"))
+				}
+
+				return nodes
+			},
+			apply: func(dll *DoubleLinkedList[string], n []*Node[string]) {
+				for _, node := range n {
+					dll.Delete(node)
+				}
+			},
+			check: func(dll *DoubleLinkedList[string]) bool {
+				if dll.Len() != 0 {
+					return false
+				}
+
+				return reflect.DeepEqual(dll.toArray(), []string{})
+			},
+		},
+		{
+			name: "Can delete the head",
+			before: func(dll *DoubleLinkedList[string]) []*Node[string] {
+				nodes := append([]*Node[string]{}, dll.Insert("One"), dll.Insert("Two"))
+
+				return nodes[0:1]
+			},
+			apply: func(dll *DoubleLinkedList[string], n []*Node[string]) {
+				for _, node := range n {
+					dll.Delete(node)
+				}
+			},
+			check: func(dll *DoubleLinkedList[string]) bool {
+				if dll.Len() != 1 {
+					return false
+				}
+
+				return reflect.DeepEqual(dll.toArray(), []string{"Two"})
+			},
+		},
+		{
+			name: "Can delete the tail",
+			before: func(dll *DoubleLinkedList[string]) []*Node[string] {
+				nodes := append([]*Node[string]{}, dll.Insert("One"), dll.Insert("Two"))
+
+				return nodes[1:]
+			},
+			apply: func(dll *DoubleLinkedList[string], n []*Node[string]) {
+				for _, node := range n {
+					dll.Delete(node)
+				}
+			},
+			check: func(dll *DoubleLinkedList[string]) bool {
+				if dll.Len() != 1 {
+					return false
+				}
+
+				return reflect.DeepEqual(dll.toArray(), []string{"One"})
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			list := NewDoubleLinkedList[string]()
+
+			tt.apply(list, tt.before(list))
+
+			if rv := tt.check(list); rv != true {
+				t.Errorf("check DoubleLinkedList.Delete() = %v, want true", rv)
+			}
+		})
+	}
+}
+
+func TestShift(t *testing.T) {
```

Comment:
I'm not sure why shifting is a functionality? But if it's required, then does it work on an empty list?

Reply:
It works on an empty list, it just returns `nil` in that case :) We can also return a `error` here, but `nil` seems more Go-like. See https://github.com/loopholelabs/common/pull/1#discussion_r918784352 for a description of why shifting is necessary.